File: electron/main/ipc/path.ts
import { ipcMain, BrowserWindow } from "electron"
import * as fs from "fs"
import * as path from "path"
import * as os from "os"
import fuzzysort from "fuzzysort"

interface PathEntry {
  name: string
  path: string
  isDir: boolean
}

interface PathSearchResult {
  entries: PathEntry[]
  error?: string
}

export function ipcPathHandler(_win: BrowserWindow) {
  ipcMain.handle("path:search", async (_event, searchPath: string): Promise<PathSearchResult> => {
    try {
      // Expand ~ to home directory
      let normalizedPath = searchPath
      if (normalizedPath.startsWith("~")) {
        normalizedPath = path.join(os.homedir(), normalizedPath.slice(1))
      }

      // Determine parent directory and prefix for filtering
      let dirToSearch: string
      let filterPrefix: string

      if (normalizedPath.endsWith(path.sep) || normalizedPath.endsWith("/")) {
        // User typed a complete directory path, list its contents
        dirToSearch = normalizedPath
        filterPrefix = ""
      } else {
        // User is typing a partial name, list parent and filter
        dirToSearch = path.dirname(normalizedPath)
        filterPrefix = path.basename(normalizedPath).toLowerCase()
      }

      // Check if directory exists
      if (!fs.existsSync(dirToSearch)) {
        return { entries: [], error: "Directory not found" }
      }

      const stat = fs.statSync(dirToSearch)
      if (!stat.isDirectory()) {
        return { entries: [], error: "Not a directory" }
      }

      // Read directory contents
      const items = fs.readdirSync(dirToSearch, { withFileTypes: true })

      // Filter and map entries
      const entries: PathEntry[] = items
        .filter(item => {
          // Hide dotfiles unless user is searching for them (filter starts with .)
          if (item.name.startsWith('.') && !filterPrefix.startsWith('.')) {
            return false
          }

          // Filter by prefix if provided
          if (filterPrefix) {
            return item.name.toLowerCase().startsWith(filterPrefix)
          }
          return true
        })
        .map(item => ({
          name: item.name,
          path: path.join(dirToSearch, item.name),
          isDir: item.isDirectory()
        }))
        // Sort: directories first, then by name
        .sort((a, b) => {
          if (a.isDir !== b.isDir) {
            return a.isDir ? -1 : 1
          }
          return a.name.localeCompare(b.name)
        })
        // Limit to 20 entries
        .slice(0, 20)

      return { entries }
    } catch (err: unknown) {
      const message = err instanceof Error ? err.message : "Unknown error"
      return { entries: [], error: message }
    }
  })

  ipcMain.handle("path:fuzzy-search", async (_event, basePath: string, query: string): Promise<PathSearchResult> => {
    try {
      // Expand ~ to home directory
      let normalizedPath = basePath
      if (normalizedPath.startsWith("~")) {
        normalizedPath = path.join(os.homedir(), normalizedPath.slice(1))
      }

      // Check if directory exists
      if (!fs.existsSync(normalizedPath)) {
        return { entries: [], error: "Directory not found" }
      }

      const stat = fs.statSync(normalizedPath)
      if (!stat.isDirectory()) {
        return { entries: [], error: "Not a directory" }
      }

      // Recursively collect all files (with depth limit)
      const allPaths: { relativePath: string; fullPath: string; isDir: boolean }[] = []
      const maxDepth = 5

      function walkDir(dir: string, depth: number, baseDir: string) {
        if (depth > maxDepth) return

        try {
          const items = fs.readdirSync(dir, { withFileTypes: true })
          for (const item of items) {
            // Skip hidden files unless query starts with .
            if (item.name.startsWith('.') && !query.startsWith('.')) {
              continue
            }

            const fullPath = path.join(dir, item.name)
            const relativePath = path.relative(baseDir, fullPath)

            allPaths.push({
              relativePath,
              fullPath,
              isDir: item.isDirectory()
            })

            // Recurse into directories
            if (item.isDirectory()) {
              walkDir(fullPath, depth + 1, baseDir)
            }
          }
        } catch {
          // Ignore permission errors
        }
      }

      walkDir(normalizedPath, 0, normalizedPath)

      // Fuzzy search using fuzzysort
      const results = fuzzysort.go(query, allPaths, {
        key: "relativePath",
        limit: 20,
        threshold: -10000
      })

      // Map results to PathEntry
      const entries: PathEntry[] = results.map(result => ({
        name: result.obj.relativePath,
        path: result.obj.fullPath,
        isDir: result.obj.isDir
      }))

      return { entries }
    } catch (err: unknown) {
      const message = err instanceof Error ? err.message : "Unknown error"
      return { entries: [], error: message }
    }
  })
}

File: src-tauri/src/command/path.rs
use serde::{Deserialize, Serialize};
use std::path::PathBuf;

#[derive(Debug, Serialize, Deserialize)]
pub struct PathEntry {
    pub name: String,
    pub path: String,
    #[serde(rename = "isDir")]
    pub is_dir: bool,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct PathSearchResult {
    pub entries: Vec<PathEntry>,
    pub error: Option<String>,
}

#[tauri::command]
pub async fn path_search(path: String) -> PathSearchResult {
    match search_path_impl(&path) {
        Ok(entries) => PathSearchResult {
            entries,
            error: None,
        },
        Err(e) => PathSearchResult {
            entries: vec![],
            error: Some(e.to_string()),
        },
    }
}

fn search_path_impl(search_path: &str) -> Result<Vec<PathEntry>, Box<dyn std::error::Error>> {
    // Expand ~ to home directory
    let normalized_path = if search_path.starts_with('~') {
        let home = dirs::home_dir().ok_or("Could not find home directory")?;
        home.join(&search_path[1..])
    } else {
        PathBuf::from(search_path)
    };

    // Determine directory to search and filter prefix
    let (dir_to_search, filter_prefix): (PathBuf, String) =
        if search_path.ends_with('/') || search_path.ends_with('\\') {
            (normalized_path.clone(), String::new())
        } else {
            let parent = normalized_path
                .parent()
                .ok_or("Invalid path")?
                .to_path_buf();
            let prefix = normalized_path
                .file_name()
                .map(|s| s.to_string_lossy().to_lowercase())
                .unwrap_or_default();
            (parent, prefix)
        };

    // Check if directory exists
    if !dir_to_search.exists() || !dir_to_search.is_dir() {
        return Ok(vec![]);
    }

    // Read directory contents
    let mut entries: Vec<PathEntry> = std::fs::read_dir(&dir_to_search)?
        .filter_map(|entry| entry.ok())
        .filter(|entry| {
            let name = entry.file_name().to_string_lossy().to_string();
            let name_lower = name.to_lowercase();

            // Hide dotfiles unless user is searching for them (filter starts with .)
            if name.starts_with('.') && !filter_prefix.starts_with('.') {
                return false;
            }

            if filter_prefix.is_empty() {
                return true;
            }
            name_lower.starts_with(&filter_prefix)
        })
        .map(|entry| {
            let metadata = entry.metadata().ok();
            PathEntry {
                name: entry.file_name().to_string_lossy().to_string(),
                path: entry.path().to_string_lossy().to_string(),
                is_dir: metadata.map(|m| m.is_dir()).unwrap_or(false),
            }
        })
        .collect();

    // Sort: directories first, then by name
    entries.sort_by(|a, b| match (a.is_dir, b.is_dir) {
        (true, false) => std::cmp::Ordering::Less,
        (false, true) => std::cmp::Ordering::Greater,
        _ => a.name.to_lowercase().cmp(&b.name.to_lowercase()),
    });

    // Limit to 20 entries
    entries.truncate(20);

    Ok(entries)
}
